### PR TITLE
Dump ast in dot format

### PR DESCRIPTION
**Pull request by taiyang-li**
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
``` sql
) explain ast graph = 1 select * from system.parts;

EXPLAIN AST graph = 1
SELECT *
FROM system.parts

Query id: bf0b8ea6-e65a-4697-b956-4705e6b17806

┌─explain────────────────────────────────────────────────────────────────┐
│ digraph {                                                              │
│     rankdir="UD";                                                      │
│     n135497629874712[label="SelectWithUnionQuery (children 1)"];       │
│     n135325831467992[label="ExpressionList (children 1)"];             │
│     n135377371300248[label="SelectQuery (children 2)"];                │
│     n135325832847912[label="ExpressionList (children 1)"];             │
│     n135325832772952[label="Asterisk"];                                │
│     n135325832847912 -> n135325832772952;                              │
│     n135325831467912[label="TablesInSelectQuery (children 1)"];        │
│     n135377371300376[label="TablesInSelectQueryElement (children 1)"]; │
│     n135411732583032[label="TableExpression (children 1)"];            │
│     n135446091858456[label="TableIdentifier system.parts"];            │
│     n135411732583032 -> n135446091858456;                              │
│     n135377371300376 -> n135411732583032;                              │
│     n135325831467912 -> n135377371300376;                              │
│     n135377371300248 -> n135325832847912;                              │
│     n135377371300248 -> n135325831467912;                              │
│     n135325831467992 -> n135377371300248;                              │
│     n135497629874712 -> n135325831467992;                              │
│ }                                                                      │
└────────────────────────────────────────────────────────────────────────┘
```

Looks like this: 
image

### CODE REVIEW

File: src/Parsers/DumpASTNode.h [hunk 1]
```diff
@@ -86,6 +86,75 @@ inline void dumpAST(const IAST & ast, WriteBuffer & ostr, DumpASTNode * parent =
...
+{
+public:
+    DumpASTNodeInDotFormat(const IAST & ast_, WriteBuffer * ostr_, bool root_ = true, const char * label_ = nullptr)
+        : ast(ast_), ostr(ostr_), root(root_), label(label_)
+    {
+        if (!ostr)
+            return;
+
+        if (root)
+            (*ostr) << "digraph " << (label ? String(label) : "") << "{\n    rankdir=\"UD\";\n";
```
Comment (vdimir) [inline src/Parsers/DumpASTNode.h:99]: To support it in `play.html` we need to have eight just "digraph" on first line, like https://github.com/ClickHouse/ClickHouse/blob/master/src/QueryPipeline/printPipeline.h#L17

or update `play.html`, use condition `response.data[0][0].startsWith("digraph")`

https://github.com/ClickHouse/ClickHouse/blob/master/programs/server/play.html#L579

**Review by vdimir — APPROVED** [review]
Thank you for your changes!

LGTM, just it worth to add some test

**Comment by taiyang-li** [comment]
@vdimir I thought about adding a test for `explain ast graph = 1`. But it's output is not deterministic, because name of nodes in dot script is generated from memory address of AST objects, just like `explain pipeline graph = 1 `

File: src/Parsers/DumpASTNode.h (hunk 1 above)
Comment (taiyang-li) [reply src/Parsers/DumpASTNode.h:99]: I think latter is better.

**Comment by taiyang-li** [comment]
Failed tests irrelavent with this PR.

**Comment by vdimir** [comment]
> @vdimir I thought about adding a test for `explain ast graph = 1`. But it's output is not deterministic, because name of nodes in dot script is generated from memory address of AST objects, just like `explain pipeline graph = 1 `. That is why `explain pipeline graph = 1` also don't have tests I guess.
> 
> Do you have any ideas? Thanks

Maybe we can add graphviz into our stateless tests image and test that generated `dot` is valid (both for pipeline and ast since we already have two places where it's used). If package size is small enough.

**Pull request merged**